### PR TITLE
Shuttle navigation computers ignore the shuttle when placing a new location

### PR DESCRIPTION
**Pull request by Cruix**
This makes it much easier to make slight adjustments to the shuttle's position.
Closes #31721 

:cl:
add: Shuttle navigation computers can now place new transit locations over the shuttle's current position.
/:cl:

### CODE REVIEW

File: code/modules/shuttle/navigation_computer.dm [hunk 1]
```diff
@@ -8,8 +8,10 @@
 	var/shuttlePortId = ""
 	var/shuttlePortName = ""
 	var/list/jumpto_ports = list() //hashset of ports to jump to and ignore for collision purposes
-	var/list/blacklisted_turfs
-	var/obj/docking_port/stationary/my_port
+	var/list/blacklisted_turfs //turfs we cannot go on, by default, any turf covered by a docking port that we did not place and cannot jump to
+	var/obj/docking_port/stationary/my_port //the custom docking port placed by this console
+	var/obj/docking_port/mobile/shuttle_port //the mobile docking port of the connected shuttle
+	var/list/shuttle_port_areas = list() //hashset of areas owned by shuttle_port
```
Comment (ninjanomnom) [inline code/modules/shuttle/navigation_computer.dm]: This list is already stores in mobile docks

File: code/modules/shuttle/navigation_computer.dm [hunk 2]
```diff
@@ -31,21 +33,23 @@
...
-	if(QDELETED(M))
+	shuttle_port = SSshuttle.getShuttle(shuttleId)
+	if(QDELETED(shuttle_port))
+		shuttle_port = null
 		return
+
+	for(var/V in shuttle_port.shuttle_areas)
+		var/area/A = V
+		if(!QDELETED(A))
+			shuttle_port_areas[A] = 1
```
Comment (ninjanomnom) [inline code/modules/shuttle/navigation_computer.dm]: Making this extra list costs more than just letting it check in the main loop

File: code/modules/shuttle/on_move.dm [hunk 3]
```diff
@@ -330,10 +330,11 @@ All ShuttleMove procs go here
 /atom/movable/light/onShuttleMove(turf/newT, turf/oldT, rotation, list/movement_force, move_dir, old_dock)
 	return FALSE
 
-/obj/docking_port/stationary/onShuttleMove(turf/newT, turf/oldT, rotation, list/movement_force, move_dir, old_dock)
+/obj/docking_port/stationary/onShuttleMove(turf/newT, turf/oldT, rotation, list/movement_force, move_dir, obj/docking_port/stationary/old_dock)
 	if(old_dock == src) //Don't move the dock we're leaving
 		return FALSE
-
+	if(old_dock && !old_dock.can_move_docking_ports_from)
+		return FALSE
```
Comment (ninjanomnom) [inline code/modules/shuttle/on_move.dm]: Instead of this make a var for stationary docks that disallows movement entirely

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by ninjanomnom** [comment]
Combining pr commits like this is a bad idea.

File: code/modules/shuttle/navigation_computer.dm (hunk 1 above)
Comment (Cruix) [reply code/modules/shuttle/navigation_computer.dm]: I did not realize that was a hashset as well.

File: code/modules/shuttle/navigation_computer.dm (hunk 2 above)
Comment (Cruix) [reply code/modules/shuttle/navigation_computer.dm]: Making this list is done once, when the eye is created. The check is done on every turf to be checked every time the eye moves, which is considerably more expensive.

File: code/modules/shuttle/on_move.dm (hunk 3 above)
Comment (Cruix) [reply code/modules/shuttle/on_move.dm]: Stationary docking ports could not move in the past, but this was changed at some point in the shuttle code rewrites. I did it this way to preserve the current behavior.

**Comment by Cruix** [comment]
Combining commits from different branches is a core function of Git.

File: code/modules/shuttle/on_move.dm (hunk 3 above)
Comment (ninjanomnom) [reply code/modules/shuttle/on_move.dm]: The behaviour was changed intentionally, I'm the main one handling docking changes and issues atm. A var for preventing individual movement is better here than stationary docks preventing all movement of other stationary dock.

File: code/modules/shuttle/navigation_computer.dm (hunk 2 above)
Comment (ninjanomnom) [reply code/modules/shuttle/navigation_computer.dm]: They *both* only get checked in CreateEye

**Comment by ninjanomnom** [comment]
Just because it's something you can do in git doesn't mean it's a good idea to use it for pull requests.

You should get in the irc chat to talk with me about these changes.

File: code/modules/shuttle/on_move.dm (hunk 3 above)
Comment (Cruix) [reply code/modules/shuttle/on_move.dm]: The only docking port with can_move_docking_ports_from set to false is the custom one placed by the shuttle docking computer. All other types of docking ports will behave exactly the same way that they do now.

File: code/modules/shuttle/navigation_computer.dm (hunk 2 above)
Comment (Cruix) [reply code/modules/shuttle/navigation_computer.dm]: https://github.com/tgstation/tgstation/pull/31722/files#diff-23c148e6e1bfb5037db9d05f3a02c9fdR129
This is the check that I am referring to.

**Comment by Cruix** [comment]
I did it this way so that my pull requests would remain atomized and the fix could be merged without the feature that depends on it, without me having to wait for the fix to be merged first.

**Comment by ninjanomnom** [comment]
The problem is because of how reviews work they aren't shared across prs and it just makes having a discussion about the changes confusing. Also this pr won't be automatically updated to have the changes of the other.

**Comment by duncathan** [comment]
can't review until the other pr goes through zzz

**Pull request merged**